### ISSUE TITLE
Kills R when there are too many trees

### ISSUE THREAD

**Issue by slavarazbash** [imbs-hl/ranger#331]
ranger version: ranger R package version 0.10.1
R version: 3.4.4 (and also on R 3.5, but needs more trees to cause the problem.)
OS: Ubuntu 18.04
Amount of RAM: 16GB

**Description**
If you build a classification or probability forest with too many trees, then it will just kill R while building the model. 
5000 trees is enough to kill it. 2000 trees is not enough.

**Minimal code to reproduce the bug**
```
library(ranger)

our.data <- data.frame(y = as.factor(ifelse(rbinom(50000,1,.5) == 1, "One", "Two")),
						x_one = rnorm(50000), x_two = rnorm(50000), x_three = rnorm(50000),
						x_four = as.factor(rpois(50000, 9)), x_five = as.factor(rpois(50000, 10)), x_six = rbinom(50000,1,.3)
						)

our.model <- ranger(dependent.variable.name="y", data=our.data, num.trees=5000)

```

**Additional Insights**
If you reduce sample.fraction, then you need to increase num.trees to hit this bug.
So the following will work fine:
`our.model <- ranger(dependent.variable.name="y", data=our.data, num.trees=5000, sample.fraction=1/1000)`
The following will kill R:
`our.model <- ranger(dependent.variable.name="y", data=our.data, num.trees=20000, sample.fraction=1/1000)`
The difference is num.trees.

**Update**
Using `top -H`, I watched the memory climb to ~89% and then the process is killed. Memory usage is growing with the number of trees built.

**Comment by mnwright** [comment]
Strange. Your example (5000 trees and default `sample.fraction`) is running perfectly fine on my Mac (R 3.4.0, 8GB) and on my Windows machine at work (R 3.4.2, 8GB). Both are using 5-6 GB while tree growing with 4 cores. My Ubuntu VM has only 2 GB and the process was killed at about 50% growing (expected). 

How many cores are you using?

Btw., the final model takes about 3.8 GB (just to save the splits), so it won't be possible to run on my VM but it should run on your machine.

**Comment by rgayler** [comment]
I *may* be having the same problem (either that or it's a hardware fault in RAM).

I am running Ubuntu 18.04, 64GB RAM, AMD processor with 16 cores / 32 threads.

The call is essentially:
```
  ranger(form, data = df, num.trees = 4000, min.node.size=200, 
         probability = TRUE, importance=importance, splitrule = "extratrees")
```
The data is ~120k rows, 300 columns.

Total memory in use by RStudio and the system was less than 6GB.

This actually ran a couple of times, but has also segfaulted, crashed R, and crashed RStudio. I suspect the randomness in outcome is from the random seeds.

I have also entertained that it is a hardware fault provoked by the heavy load, but swapping the RAM modules around makes no difference, and I haven't been able to provoke any problems with a stress testing tool.

**Comment by mnwright** [comment]
Could you try to simulate data close to your actual data? Somehow we'll have to reproduce the error to find the problem.

**Comment by slavarazbash** [comment]
@mnwright On my machine, I found that there was another process using a non-trivial amount of RAM. I killed it and now the 5000 trees example runs. What threw me off was that it would die just at the end of the model build - when CPU usage was already dropping. I noticed that there is a bump RAM usage at the end of the model build - that was pushing it over the edge.

To answer your question: 4 cores with hyper-threading.
